packages/core/src/common/uri.ts

```diff
@@ -51,7 +51,7 @@ export default class URI {
     get allLocations(): URI[] {
         const locations = [];
         let location: URI = this;
-        while (!location.path.isRoot) {
+        while (!location.path.isRoot && location.path.hasDir) {
```

Comment:
I think `location.path.hasDir` should not be necessary.

Since we cannot remove `undefined` for `path.dir`, can we remove an empty path instead? It would be aligned when with Node.js behaviour as it should be, i.e.
```js
const path = require('path');
console.log(path.parse('foo'));
```

prints:
```
gitpod /workspace/theia $ node test.js 
{ root: '', dir: '', base: 'foo', ext: '', name: 'foo' }
```

Reply:
Do you mean that `Path#dir` should return an empty string instead of  `this` when there is no dir? If you mean that, then I would totally agree with you. In my opinion, that would be much more intuitive. I just wonder what the reason for returning `this` was in the first place … and if changing this would break something elsewhere.

Reply:
I meant an empty path, i.e. something like `new Path('')`.

My concern is that a loop like here should always terminate for any URI just by checking `path.isRoot`.

Reply:
Sorry for the confusion, just spotted that i typed `remove` instead of `return` in the original comment.

Reply:
And I meant empty path instead of empty string. :-D

I totally agree with you that an empty path instead of `this` makes more sense as return value. The same goes in my opinion for `URI#parent`, by the way. But this does not terminates the loop above. A path that is not absolute does not have a root at all and therefore asking for `isRoot` is not right here. Or am I wrong?

But the loop condition above can be abbreviated by just checking for `hasDir` because root never has a dir.

Reply:
yes, I think you are right, let's go with this solution it does not seem to be breaking